src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,175 @@
+        int i = 0;
+        while (currentNode != null) {
+            if (i == index) {
+                Node<T> node = new Node<T>(currentNode.prev, value, currentNode);
+                if (index == 0) {
+                    firstNode = node;
+                } else {
+                    currentNode.prev.next = node;
+                }
+                currentNode.prev = node;
+                size++;
+                return;
+            }
+            currentNode = currentNode.next;
+            i++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
+        checkIndexInBounds(index);
+        Node<T> currentNode = firstNode;
+        int i = 0;
+        while (currentNode != null) {
+            if (i == index) {
+                return currentNode.value;
+            }
+            currentNode = currentNode.next;
+            i++;
+        }
         return null;
     }
 
     @Override
     public T set(T value, int index) {
+        checkIndexInBounds(index);
+        Node<T> currentNode = firstNode;
+        int i = 0;
+        while (currentNode != null) {
+            if (i == index) {
+                T oldValue = currentNode.value;
+                currentNode.value = value;
+                return oldValue;
+            }
+            currentNode = currentNode.next;
+            i++;
+        }
         return null;
     }
 
     @Override
     public T remove(int index) {
+        checkIndexInBounds(index);
+        if (index == 0) {
+            if (firstNode.next != null) {
+                firstNode.next.prev = null;
+            }
+            T oldValue = firstNode.value;
+            firstNode = firstNode.next;
+            size--;
+            return oldValue;
+        }
+        if (index == size - 1) {
+            lastNode.prev.next = null;
+            T oldValue = lastNode.value;
+            lastNode = lastNode.prev;
+            size--;
+            return oldValue;
+        }
+        Node<T> currentNode = firstNode.next;
+        int i = 1;
+        do {
+            if (i == index) {
+                currentNode.prev.next = currentNode.next;
+                currentNode.next.prev = currentNode.prev;
+                T oldValue = currentNode.value;
+                currentNode.value = currentNode.next.value;
+                size--;
+                return oldValue;
+            }
+            currentNode = currentNode.next;
+            i++;
+        } while (currentNode != lastNode);
         return null;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = firstNode;
+        int i = 0;
+        while (currentNode != null) {
+            if ((currentNode.value != null && currentNode.value.equals(object))
+                    || currentNode.value == object) {
+                remove(i);
```

Comment:
here it will be enough to use the new unlink () method, in our remove (index) method there is much more logic than is necessary in this method
By moving the actions to redistribute prev and next between nodes into the unlink method, it will be better.